CDR limitations: share of FE and share of expenses for net negative emissions in GDP

tabeado opened the pull request:
## Purpose of this PR
This PR introduces the possibility to limit CDR based on energy and monetary resources:
a) **share of final energy type demand in total demand of that FE type in a region** via the switch `cm_33_maxFeShare`. 
- It determines an upper bound for each final energy carrier based on the previous iteration's total final energy of that type. This avoids that the model increases other sector's FE demand in response to the limitation. 
- We aggregate similar FE carriers via the mapping `entyFe2FeType`. This means for example, that the demand for electricity (`feels`) by `dac + weathering + oae_ng + oae_el ` has to be less than X% of `feels`**+`feelt`**.

b) **Limit expenses on net negative emissions:** The new variable `v_NetNegEmi_expenses` calculates the expenses for net negative emissions as the effective carbon price for net negative emissions times the total net negative emissions. The switch `c_GDP_netNegCDR_maxShare` determines the maximum share thereof in GDP that is permissible by setting an upper bound on `v_NetNegEmi_expenses`.

Both `cm_33_maxFeShare` and `c_GDP_netNegCDR_maxShare`  are set to 1 as default, thus not affecting default runs. 

The switches allow to address the [issue ](https://github.com/orgs/remindmodel/projects/15/views/2?pane=issue&itemId=92689490&issue=remindmodel%7Cdevelopment_issues%7C257) of regional concentration of certain technologies if desired in a scenario. 

## Type of change
- [x] New feature 
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the [coding etiquette](https://github.com/remindmodel/remind/blob/develop/main.gms#L80)
- [x] I performed a self-review of my own code
- [x] I explained my changes within the PR, particularly in hard-to-understand areas
- [x] I checked that the [in-code documentation](https://github.com/remindmodel/remind/blob/develop/main.gms#L120) is up-to-date
- [ ] I adjusted the reporting in [`remind2`](https://github.com/pik-piam/remind2) where it was needed
- [ ] I adjusted `forbiddenColumnNames` in [readCheckScenarioConfig.R](https://github.com/remindmodel/remind/blob/develop/scripts/start/readCheckScenarioConfig.R) in case the PR leads to deprecated switches
- [ ] I checked the `log.txt` file of my runs for newly introduced summation, fixing or variable name errors
- [x] All automated model tests pass (`FAIL 0` in the output of `make test`)
- [ ] The changelog `CHANGELOG.md` [has been updated correctly](https://gitlab.pik-potsdam.de/rse/rsewiki/-/wikis/Standards-for-Writing-a-Changelog)

## Further information (optional):

* Runs with these changes are here: /p/tmp/tabeado/CDRregiLimitsF/remind
* Comparison of results (what changes by this PR?):

LaviniaBaumstark commented on core/datainput.gms:
```diff
@@ -128,6 +128,16 @@ pm_shfe_lo(ttot,regi,entyFe,sector)=0;
+pm_shfetot_up(t,regi,entyFe,sector)=0;
+
+*** limit on expenses for net negative emissions as share of GDP
+parameter f_GDP_NetNeg_share(ext_regi) "TEXT" / %c_GDP_netNegCDR_maxShare% /; 
```
is this file already in the new input data?

tabeado replied:
it is reading a switch, not a new input file. I did it analogous to cm_emiMktTarget_tolerance and f47_emiMktTarget_tolerance [here](https://github.com/remindmodel/remind/blob/5a5c2c8f03282053150f5c69882af8175e6e308c/modules/47_regipol/regiCarbonPrice/datainput.gms#L62)
I didn't check the f_ meaning, will change to p_

LaviniaBaumstark commented on core/equations.gms:
```diff
@@ -1251,4 +1251,41 @@ q_fossilSolidsLimitReg(ttot,regi,entySe,entyFe,sector,emiMkt)$(limitSolidsFossil
 $endif.limitSolidsFossilRegi
 
+***---------------------------------------------------------------------------
+*' Limit the amount of FE for CDR to a given fraction of total FE 
```
could those equations maybe go into a (new) module instead of core?

LaviniaBaumstark commented on core/postsolve.gms:
```diff
@@ -77,6 +77,12 @@ loop(ttot$(ttot.val ge 2005),
 pm_GDPGross(tall,regi)$(tall.val ge 2150) = pm_GDPGross("2149",regi); 
 
+*** update bound on spending on net negative emissions 
+v_NetNegEmi_expenses.up(t,regi) = p_GDP_NetNeg_share(regi) * pm_gdp_gdx(t,regi);
```
this will have no effect here or is this variable used later in postsolve?

tabeado replied:
The upper bound on the variable is updated in postsolve based on the actual gdp from the last iteration. 
It then takes effect as upper bound for the variable calculation in equations in the next iteration.

This is mainly to avoid that REMIND can increase other FE uses within an iteration to maximize CDR-FE

LaviniaBaumstark commented on core/postsolve.gms:
```diff
@@ -77,6 +77,12 @@ loop(ttot$(ttot.val ge 2005),
+
+*** update bound on final energy for a sector
+p_FE_limit(t,regi,entyFE,sector)$pm_shfetot_up(t,regi,entyFe,sector) = pm_shfetot_up(t,regi,entyFE,sector) * sum(entyFe2FeType(entyFe2,entyFe), v_FE_total.l(t,regi,entyFe2));
+v_FEsector_total.up(t,regi,entyFe,sector)$pm_shfetot_up(t,regi,entyFe,sector) = p_FE_limit(t,regi,entyFE,sector);
```
same for this variable

tabeado commented:
I moved everything to 33_cdr and renamed the variables accordingly. 
There is visible difference between test runs before and after the move (middle scenario: before, right scenario: after the move)
![image](https://github.com/user-attachments/assets/5f85fb66-585f-42e6-bafa-f6cbcf570b5e)
![image](https://github.com/user-attachments/assets/baa7fff2-b3ce-42b2-840e-036009f1e5c3)

![image](https://github.com/user-attachments/assets/8e1360bf-bb37-4e64-9527-b0312e8e0aad)

test runs + compareScen are here: `/p/tmp/tabeado/CDRregiLimitsF/remind`

I am wondering whether this could be connected to the [missing enforcement of same fe-shares between sectors for CDR](https://github.com/orgs/remindmodel/projects/16?pane=issue&itemId=90808451&issue=remindmodel%7Cdevelopment_issues%7C416), which gives REMIND random flexibility to move things around?

tabeado commented:
Differences are most likely because of the unfortunate [update with develop](https://github.com/remindmodel/remind/pull/1971/commits/687f7dc03e054229cbf99a6c8a9aa9e432900719)

The limitations are actually binding in the same regions, for example electricity in MEA (3% limit)
![image](https://github.com/user-attachments/assets/d9e78af1-7106-4708-8430-9e646ea17d58)

LaviniaBaumstark reviewed (commented):
please alos run make test with teh final version

lecfab commented:
> Both cm_33_maxFeShare and c_GDP_netNegCDR_maxShare are set to 1 as default, thus not affecting default runs.

Hey!
After reading this, I was convinced that there was no default CDR limit. But it seems to be rather set to 10%?
would it be possible to edit the message in case someone else ends up here? =)

tabeado commented:
Thanks for pointing this out, @lecfab - we changed the default after I had drafted the initial message. The text has been adjusted. 

Unfortunately, there is no reference for 10%. There is a study by [Beerling et al (2020)](https://www.nature.com/articles/s41586-020-2448-9 ) limiting Enhanced Weathering potential based on 3% of energy demand.
